Keyboard closes after typing one character in gif search field on x.com

lilyspiniolas opened the pull request:
#### bdad3ac0ce7807eb1ed9fb8923a8ecb7dc58728d
<pre>
Keyboard closes after typing one character in gif search field on x.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=298304">https://bugs.webkit.org/show_bug.cgi?id=298304</a>
<a href="https://rdar.apple.com/162808064">rdar://162808064</a>

Reviewed by Abrar Rahman Protyasha.

If the keyboard is hidden and an element is programatically focused during
the same presentation update, allow the keyboard to show again.

Track explicit user dismissals of the keyboard using `_editingEndedWithReason`,
and only restore the keyboard if `_editingEndedWithReason` is false.

These changes were originally intoduced in 299615@main but broke async image
decoding for first tile paint and were reverted in 301564@main. This is because
`WebPageProxy::callAfterNextPresentationUpdate()` was being called every time
`_hideKeyboard` was called, which forces a synchronous image decode. This time
around, we pass `editableChanged` to `_hideKeyboard` so that we only call
`WebPageProxy::callAfterNextPresentationUpdate()` if the editing state actually
changed. Thanks to Wenson Hsieh for this fix.

With these changes, layout test
tap-to-change-selection-after-accepting-inline-prediction.html began to fail
due to it using `&lt;body contenteditable&gt;` as the focused element. When
the body was blurred, focus was restored to the body again, and the keyboard
remained open as a result. The test has been updated to use `&lt;div contenteditable&gt;`
instead so that the keyboard is correctly dismissed. This issue appears to only
affect the simulator; programmatic keyboard dismissal on a contenteditable body
using `blur()` still works for both software and hardware keyboard.

Test: fast/forms/ios/keyboard-restoration-after-element-replacement.html
* LayoutTests/editing/selection/ios/tap-to-change-selection-after-accepting-inline-prediction.html:
* LayoutTests/fast/forms/ios/keyboard-restoration-after-element-replacement-expected.txt: Added.
* LayoutTests/fast/forms/ios/keyboard-restoration-after-element-replacement.html: Added.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView endEditingAndUpdateFocusAppearanceWithReason:]):
(-[WKContentView _didCommitLoadForMainFrame]):
(-[WKContentView _hideKeyboard:]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _elementDidBlur]):
(-[WKContentView _hideKeyboard]): Deleted.

Canonical link: <a href="https://commits.webkit.org/301887@main">https://commits.webkit.org/301887@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1c48247914d097eac30df20f4297b4919e51737

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127390 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/47038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38175 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/134454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129262 "Build is in progress. Recent messages:") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/47654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/55561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/134454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130338 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/47654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/114098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/134454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/47654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/77836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/47654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/32619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/136937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/54049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/55561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/136937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/54560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/110447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/136937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19920 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/53986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/60073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/54979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/editing/selection/ios/tap-to-change-selection-after-accepting-inline-prediction.html:
```diff
@@ -76,16 +76,18 @@
 
     console.log("8. Dismissing keyboard");
-    document.body.blur();
+    document.getElementById("editor").blur();
```
hmm does this mean that the keyboard can no longer be programmatically dismissed on a `<body contenteditable>`?

lilyspiniolas replied:
The keyboard can still be programmatically dismissed; this seems to be an issue unique to the simulator. I'll verify with a hardware keyboard shortly.

pxlcoder replied:
(if that ends up being the case, please mention in the commit message 🙂)

lilyspiniolas replied:
I verified with a hardware keyboard as well, will update commit message.

pxlcoder reviewed (commented):
commit message typo: `everytime` -> `every time`

lilyspiniolas committed:
Keyboard closes after typing one character in gif search field on x.com
https://bugs.webkit.org/show_bug.cgi?id=298304
rdar://162808064

Reviewed by Abrar Rahman Protyasha.

If the keyboard is hidden and an element is programatically focused during
the same presentation update, allow the keyboard to show again.

Track explicit user dismissals of the keyboard using `_editingEndedWithReason`,
and only restore the keyboard if `_editingEndedWithReason` is false.

These changes were originally intoduced in 299615@main but broke async image
decoding for first tile paint and were reverted in 301564@main. This is because
`WebPageProxy::callAfterNextPresentationUpdate()` was being called every time
`_hideKeyboard` was called, which forces a synchronous image decode. This time
around, we pass `editableChanged` to `_hideKeyboard` so that we only call
`WebPageProxy::callAfterNextPresentationUpdate()` if the editing state actually
changed. Thanks to Wenson Hsieh for this fix.

With these changes, layout test
tap-to-change-selection-after-accepting-inline-prediction.html began to fail
due to it using `<body contenteditable>` as the focused element. When
the body was blurred, focus was restored to the body again, and the keyboard
remained open as a result. The test has been updated to use `<div contenteditable>`
instead so that the keyboard is correctly dismissed. This issue appears to only
affect the simulator; programmatic keyboard dismissal on a contenteditable body
using `blur()` still works for both software and hardware keyboard.

Test: fast/forms/ios/keyboard-restoration-after-element-replacement.html
* LayoutTests/editing/selection/ios/tap-to-change-selection-after-accepting-inline-prediction.html:
* LayoutTests/fast/forms/ios/keyboard-restoration-after-element-replacement-expected.txt: Added.
* LayoutTests/fast/forms/ios/keyboard-restoration-after-element-replacement.html: Added.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView endEditingAndUpdateFocusAppearanceWithReason:]):
(-[WKContentView _didCommitLoadForMainFrame]):
(-[WKContentView _hideKeyboard:]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _elementDidBlur]):
(-[WKContentView _hideKeyboard]): Deleted.

Canonical link: https://commits.webkit.org/301887@main
